File: Lab1/Lab_1_Problem_1.m
#### WRITTEN IN OCTAVE, Minor adjustments required if to be run in matlab
# Prof. Jefferies said it was not required to do so.


n = input("Enter size of matricies\n");
A = rand(n,n);
B = rand(n,n);
startTime = cputime();
myC = MM(A,B);
myTime = cputime() - startTime;
startTime = cputime();
trueAns = A*B;
matTime = cputime() - startTime;
if(isequal(round(myC.*10000)./10000, round(trueAns.*10000)./10000))
  disp("My MM took ");
  disp(myTime);
  disp("seconds");
  disp("Matlab took");
  disp(matTime);
  disp("seconds");
endif
myTimes = [];
matTimes = [];
sizes = [];
n = 2
for i = 1:7
  A = rand(n,n);
  B = rand(n,n);
  startTime = cputime();
  myC = MM(A, B);
  myTime = cputime() - startTime;
  myTimes(i) = myTime;
  startTime = time(); %gets seconds since the epoch
  matC = A*B;
  matTime = time() - startTime;
  matTime = matTime*100000;%now uS
  matTimes(i) = matTime;
  sizes(i) = n;
  n = n*2;
endfor

figure
hold
plot(sizes, myTimes);
plot(sizes, matTimes);
title("Matrix Multiplication Octave");
xlabel("n");
ylabel("time(s)");
legend("My matrix multiplication in s", "Octave's matrix multiplication * 10^5");

File: OldLab1/PyhtonPart/Lab_1_Problem_1.py
import numpy as np
import time
import matplotlib.pyplot as plt
'''
def matrixmultiply(Ar, Ac, Br, Bc):
    M1 = np.random.random((Ar,Ac))
    M2 = np.random.random((Br,Bc))
    if(Ac!=Br):
        return (-1,0)
    P = np.zeros((Ar, Bc))
    startTime = time.time()
    for row in range(Ar):
        for col in range(Bc):
            sum_ = 0
            for i in range(Ac):
                sum_ += M1[row, i]*M2[i, col]
            P[row, col] = sum_
    myTime = time.time()-startTime
    startTime = time.time()
    trueAns = np.matmul(M1, M2)
    npTime = time.time() - startTime
    if not np.array_equal(np.round(trueAns), np.round(P)):
        return (-2, 0)
    return (myTime, npTime)
'''
def matrixmultiply(A, B):
    if(np.shape(A)[1]!=np.shape(B)[0]):
        return -1
    Ar, Ac = np.shape(A)
    Br, Bc = np.shape(B)
    C = np.zeros((Ar,Bc))
    for row in range(Ar):
        for col in range(Bc):
            sum_ = 0
            for i in range(Ac):
                sum_ += A[row, i] * B[i, col]
            C[row, col] = sum_
    return C
n = int(input("Starting Sizes\n"))
A = np.random.rand(n,n)
B = np.random.rand(n,n)
startTime = time.time()
C = matrixmultiply(A, B)
myTime = time.time() - startTime
startTime = time.time()
trueAns = np.matmul(A, B)
numpyTime = time.time()-startTime
if np.array_equal(np.round(trueAns), np.round(C)):
    print("My MM took " + str(myTime)+ " seconds")
    print("Numpys MM took "+str(numpyTime)+ " seconds")
n = 4
myTimes = []
npTimes = []
sizes = []
for i in range(7):
    A = np.random.rand(n, n)
    B = np.random.rand(n, n)
    startTime = time.time()
    C = matrixmultiply(A, B)
    myTime = time.time() - startTime
    startTime = time.time()
    trueAns = np.matmul(A, B)
    numpyTime = time.time() - startTime
    if not np.array_equal(np.round(trueAns), np.round(C)):
        continue
    myTimes.append(myTime)
    npTimes.append(numpyTime)
    sizes.append(n)
    print("finished size " + str(n))
    n = n*2

plt.plot(sizes, myTimes)
#plt.plot(sizes, npTimes)
plt.xlabel("n")
plt.ylabel("time(s)")
plt.title("My Matrix Multiplication in python")
plt.show()
plt.figure()

n = 100
myTimes = []
npTimes = []
sizes = []
for i in range(7):
    A = np.random.rand(n, n)
    B = np.random.rand(n, n)
    #startTime = time.time()
    #C = matrixmultiply(A, B)
    #myTime = time.time() - startTime
    startTime = time.time()
    trueAns = np.matmul(A, B)
    numpyTime = time.time() - startTime

    myTimes.append(myTime)
    npTimes.append(numpyTime)
    sizes.append(n)
    print("finished size " + str(n))
    n = n*2

#plt.plot(sizes, myTimes)
plt.plot(sizes, npTimes)
plt.xlabel("n")
plt.ylabel("time(s)")
plt.title("Numpys Matrix Multiplication in python")
plt.show()
plt.figure()